src/phoenix/trace/openai/instrumentor.py

```diff
@@ -0,0 +1,286 @@
+    SpanException,
+    SpanKind,
+    SpanStatusCode,
+)
+from phoenix.trace.semantic_conventions import (
+    INPUT_MIME_TYPE,
+    INPUT_VALUE,
+    LLM_FUNCTION_CALL,
+    LLM_INPUT_MESSAGES,
+    LLM_INVOCATION_PARAMETERS,
+    LLM_OUTPUT_MESSAGES,
+    LLM_TOKEN_COUNT_COMPLETION,
+    LLM_TOKEN_COUNT_PROMPT,
+    LLM_TOKEN_COUNT_TOTAL,
+    MESSAGE_CONTENT,
+    MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON,
+    MESSAGE_FUNCTION_CALL_NAME,
+    MESSAGE_NAME,
+    MESSAGE_ROLE,
+    OUTPUT_MIME_TYPE,
+    OUTPUT_VALUE,
+    MimeType,
+)
+from phoenix.trace.utils import get_stacktrace, import_package
+
+from ..tracer import Tracer
+
+if TYPE_CHECKING:
+    from openai.openai_response import OpenAIResponse
+
+
+Parameters = Mapping[str, Any]
+OpenInferenceMessage = Dict[str, str]
+
+INSTRUMENTED_ATTRIBUTE_NAME = "is_instrumented_with_openinference_tracer"
+
+
+class RequestType(Enum):
+    CHAT_COMPLETION = "chat_completion"
+    COMPLETION = "completion"
+    EMBEDDING = "embedding"
+
+
+class OpenAIInstrumentor:
+    def __init__(self, tracer: Optional[Tracer] = None) -> None:
+        """Instruments your OpenAI client to automatically create spans for each API call.
+
+        Args:
+            tracer (Optional[Tracer], optional): A tracer to record and handle spans. If not
+            provided, the default tracer will be used.
+        """
+        self._tracer = tracer or Tracer()
+
+    def instrument(self) -> None:
+        """
+        Instruments your OpenAI client.
+        """
+        openai = import_package("openai")
+        is_instrumented = hasattr(
+            openai.api_requestor.APIRequestor.request,
+            INSTRUMENTED_ATTRIBUTE_NAME,
+        )
+        if not is_instrumented:
+            openai.api_requestor.APIRequestor.request = _wrap_openai_api_requestor(
+                openai.api_requestor.APIRequestor.request, self._tracer
+            )
+            setattr(
+                openai.api_requestor.APIRequestor.request,
+                INSTRUMENTED_ATTRIBUTE_NAME,
+                True,
+            )
+
+
+def _wrap_openai_api_requestor(
+    request_fn: Callable[..., Any], tracer: Tracer
+) -> Callable[..., Any]:
+    """Wraps the OpenAI APIRequestor.request method to create spans for each API call.
+
+    Args:
+        request_fn (Callable[..., Any]): The request method on openai.api_requestor.APIRequestor.
+        tracer (Tracer): The tracer to use to create spans.
+
+    Returns:
+        Callable[..., Any]: The wrapped request method.
+    """
+
+    def wrapped(*args: Any, **kwargs: Any) -> Any:
+        call_signature = signature(request_fn)
+        bound_arguments = call_signature.bind(*args, **kwargs)
+        parameters = bound_arguments.arguments["params"]
+        is_streaming = parameters.get("stream", False)
+        url = bound_arguments.arguments["url"]
+        current_status_code = SpanStatusCode.UNSET
+        events: List[SpanEvent] = []
+        attributes: SpanAttributes = dict()
```

Comment:
I don't have an opinion on this one. I've probably seen `{}` more frequently in other codebases, and it's certainly more common in our codebase. Going to skip this one since @anticorrelator expressed the opposite preference.

Reply:
I think the empty dictionary literal is easily confused with set notation so I prefer to use the constructor instead 🙃

Reply:
technically, `dict()` is a function call and so is slower.

<img width="676" alt="Screenshot 2023-10-04 at 10 15 05 PM" src="https://github.com/Arize-ai/phoenix/assets/80478925/9317a82f-0c75-4cf2-a6cd-a0335c7c924c">